### PR TITLE
Backport upstream changes to watch cache enablement

### LINKED ISSUE #16112: Assess impact of watch cache on large production cluster (5k+ namespaces)
On a large dense cluster memory use of the api server is of primary concern (for scale) due to the watch cache keeping copies of all objects in memory.  We conducted a test on a cluster to assess the impact of disabling the watch cache.

Scenario:
* HA cluster, masters running OpenShift 3.6 (Kube 1.6+openshift patches).  ~170 nodes 5k namespaces
* Disabled watch cache on the API server of the master that the active controller was talking to
* Measured results via prometheus on that master
* Restart was at 17:00 UTC

CPU over time for that master cpu (red is apiserver, blue is etcd)

RSS for that master mem (purple is apiserver, brown is controller)

Internal go heap stats for all masters heap

No observed change to tail latency.  Still measuring impact to CPU vs number of small watches

### PR DESCRIPTION
**Pull request by smarterclayton**
Disables the watch cache for most resources by default, except those accessed by many clients. This has been shown to have minor impacts on the production workload.

Fixes #16112

### CODE REVIEW

File: pkg/cmd/server/kubernetes/master/master_config.go [hunk 1]
```diff
@@ -150,7 +148,7 @@ func BuildKubeAPIserverOptions(masterConfig configapi.MasterConfig) (*kapiserver
 	server.Etcd.StorageConfig.KeyFile = masterConfig.EtcdClientInfo.ClientCert.KeyFile
 	server.Etcd.StorageConfig.CertFile = masterConfig.EtcdClientInfo.ClientCert.CertFile
 	server.Etcd.StorageConfig.CAFile = masterConfig.EtcdClientInfo.CA
-	server.Etcd.DefaultWatchCacheSize = DefaultWatchCacheSize
+	server.Etcd.DefaultWatchCacheSize = 0
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/master/master_config.go]: This is what is setting us to "off by default", right?

File: pkg/cmd/server/kubernetes/master/master_config.go [hunk 2]
```diff
@@ -507,6 +505,20 @@ func buildKubeApiserverConfig(
 		return originLongRunningRequestRE.MatchString(r.URL.Path) || kubeLongRunningFunc(r, requestInfo)
 	}
 
+	if apiserverOptions.Etcd.EnableWatchCache {
+		glog.V(2).Infof("Initializing cache sizes based on %dMB limit", apiserverOptions.GenericServerRunOptions.TargetRAMMB)
+		sizes := cachesize.NewHeuristicWatchCacheSizes(apiserverOptions.GenericServerRunOptions.TargetRAMMB)
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/master/master_config.go]: do we set this target RAMMB to anything by default?

Comment (deads2k) [reply pkg/cmd/server/kubernetes/master/master_config.go]: > do we set this target RAMMB to anything by default?

I'm not seeing where we write a default here, which would set all the heuristic ones to zero by default, right?

File: pkg/cmd/server/kubernetes/master/master_config.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/cmd/server/kubernetes/master/master_config.go]: Correct

File: pkg/cmd/server/kubernetes/master/master_config.go (hunk 2 above)
Comment (smarterclayton) [reply pkg/cmd/server/kubernetes/master/master_config.go]: There's a "min" function on the heuristic so we always get something even at 0

**Comment by smarterclayton** [comment]
Anything else?

**Comment by smarterclayton** [comment]
ping @deads2k

**Comment by jeremyeder** [comment]
@openshift/svt

**Pull request merged**